test/zipkin/util.go

```diff
@@ -0,0 +1,135 @@
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+
+	"github.com/knative/pkg/test/logging"
+	"go.opencensus.io/trace"
+	"k8s.io/client-go/kubernetes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+
+	// ZipkinPortForwardPid environment variable name to store Process ID of the process started for zipkin port-foward.
+	ZipkinPortForwardPid = "ZIPKIN_PORTFWD_PID"
+
+	//ZipkinTraceIDHeader HTTP response header key to be used to store Zipkin Trace ID.
+	ZipkinTraceIDHeader = "ZIPKIN_TRACE_ID"
+
+	// ZipkinPort port exposed by the Zipkin Pod
+	ZipkinPort = 9411
+
+	// ZipkinTraceEndpoint port-forwarded zipkin endpoint
+	ZipkinTraceEndpoint = "http://localhost:9411/api/v2/trace/%s"
+
+	// ZipkinNamespace namespace where zipkin pod runs.
+	ZipkinNamespace = "istio-system"
+)
+
+// SetupZipkinTracing sets up zipkin tracing which involves a) Setting up port-forwarding from localhost to zipkin pod on the cluster (pid of the process doing Port-Forward is stored in an env variable).
+// b) enable AlwaysSample config for tracing.
+func SetupZipkinTracing(kubeClientset *kubernetes.Clientset) error {
+
+	logger := logging.GetContextLogger("SpoofUtil")
+	if !CheckZipkinPortAvailability() {
+		return errors.New("Zipkin port not available on the machine. Zipkin Setup failed")
+	}
+
+	zipkinPods, err := kubeClientset.CoreV1().Pods(ZipkinNamespace).List(metav1.ListOptions{LabelSelector: "app=zipkin"})
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving Zipkin pod details : %v. Zipkin Setup failed", err)
+	}
+
+	if len(zipkinPods.Items) == 0 {
+		return errors.New("No Zipkin Pod found on the cluster. Ensure monitoring is switched on for your Knative Setup. Zipkin Setup failed")
+	}
+
+	portForwardCmd := exec.Command("kubectl", "port-forward", fmt.Sprintf("--namespace=%s", ZipkinNamespace), zipkinPods.Items[0].Name, fmt.Sprintf("%d:%d", ZipkinPort, ZipkinPort))
+	err = portForwardCmd.Start()
+
+	if err != nil {
+		return fmt.Errorf("Error starting kubectl port-forward command : %v", err)
+
+	}
+
+	os.Setenv(ZipkinPortForwardPid, strconv.Itoa(portForwardCmd.Process.Pid))
+
+	// Applying AlwaysSample config to ensure we propagate zipkin header for every request made by this client.
+	trace.ApplyConfig(trace.Config{DefaultSampler: trace.AlwaysSample()})
+
+	logger.Infof("Successfully setup SpoofingClient for Zipkin Tracing")
+
+	return nil
+}
+
+// CleanupZipkinTracingSetup cleans up the Zipkin tracing setup on the machine. This involves killing the process performing port-forward.
+// Method expects the port-forwarding process to be stored in an environment variable.
+func CleanupZipkinTracingSetup() error {
+
+	logger := logging.GetContextLogger("SpoofUtil")
+	pid := os.Getenv(ZipkinPortForwardPid)
+
+	if len(pid) == 0 {
+		return fmt.Errorf("Zipking port-forward Process ID not found. CleanupZipkingTracingSetup(..) expects %s environment variable to to be pointing to Process ID perform port-forwarding", ZipkinPortForwardPid)
```

Comment:
nit: "to to" and unnecessary ".."

Reply:
Fixed.

Reply:
I'll ignore the missed "perform" because I'm sure you'll abandon this env var approach, marking as resolved. :)